File: java/mediator/ChatManagement.java
package mediator;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ChatManagement implements ChatMediator {
    private Map<String, List<User>> groups;

    public ChatManagement() {
        this.groups = new HashMap<>();
    }

    @Override
    public void sendDirectMessage(String message, User toUser, User fromUser) {
        System.out.println(fromUser.getName() + " is sending message: "
                + message + " to user: " + toUser.getName());
        toUser.receiveDirectMessage(message, fromUser);
    }

    @Override
    public void sendGroupMessage(String message, User fromUser, String toGroup) {
        System.out.println(fromUser.getName() + " is sending message: " + message + " to group: " + toGroup);
        var users = groups.get(toGroup);
        users.forEach(user -> user.receiveGroupMessage(message, fromUser, toGroup));
    }

    @Override
    public void registerUserToGroup(User user, String groupName) {
        if (!groups.containsKey(groupName)) {
            List<User> users = new ArrayList<>();
            users.add(user);
            groups.put(groupName, users);
            return;
        }
        groups.get(groupName).add(user);
    }
}

File: C#/Mediator/ChatManagement.cs
using System.Runtime.InteropServices;

namespace Mediator;

public sealed class ChatManagement : IChatMediator
{
    private readonly Dictionary<string, List<User>> _groups;

    public ChatManagement()
        => _groups = [];

    public void RegisterUserToGroup(User user, string groupName)
    {
        if (!_groups.TryGetValue(groupName, out List<User>? value))
        {
            value = [user];
            _groups.Add(groupName, value);
            return;
        }

        _groups[groupName].Add(user);
    }

    public void SendDirectMessage(string message, User toUser, User fromUser)
    {
        Console.WriteLine($"{fromUser.Name} is sending message: {message} to user: {toUser.Name}");
        toUser.ReceiveDirectMessage(message, fromUser);
    }

    public void SendGroupMessage(string message, User fromUser, string toGroup)
    {
        Console.WriteLine($"{fromUser.Name} is sending message: {message} to group: {toGroup}");
        List<User> users = _groups[toGroup];

        foreach (var user in CollectionsMarshal.AsSpan(users))
        {
            user.ReceiveGroupMessage(message, fromUser, toGroup);
        }
    }
}
